spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/ComputeTableStatsSparkAction.java

```diff
@@ -0,0 +1,164 @@
+  protected ComputeTableStatsSparkAction self() {
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    String desc = String.format("Computing stats for %s", table.name());
+    JobGroupInfo info = newJobGroupInfo("COMPUTE-TABLE-STATS", desc);
+    return withJobGroupInfo(info, this::doExecute);
+  }
+
+  private Result doExecute() {
+    if (snapshot == null) {
+      return EMPTY_RESULT;
+    }
+    LOG.info("Computing stats of {} for snapshot {}", table.name(), snapshot.snapshotId());
+    List<Blob> blobs = generateNDVBlobs();
+    StatisticsFile statisticsFile;
+    try {
+      statisticsFile = writeStatsFile(blobs);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+    table.updateStatistics().setStatistics(snapshot.snapshotId(), statisticsFile).commit();
+    return ImmutableComputeTableStats.Result.builder().statisticsFile(statisticsFile).build();
+  }
+
+  private StatisticsFile writeStatsFile(List<Blob> blobs) throws IOException {
+    LOG.info("Writing stats for table {} for snapshot {}", table.name(), snapshot.snapshotId());
+    TableOperations operations = ((HasTableOperations) table).operations();
+    String path = operations.metadataFileLocation(String.format("%s.stats", UUID.randomUUID()));
+    OutputFile outputFile = operations.io().newOutputFile(path);
+    try (PuffinWriter writer = Puffin.write(outputFile).createdBy(appIdentifier()).build()) {
+      blobs.forEach(writer::add);
+      writer.finish();
+      return new GenericStatisticsFile(
+          snapshot.snapshotId(),
+          path,
+          writer.fileSize(),
+          writer.footerSize(),
+          writer.writtenBlobsMetadata().stream()
+              .map(GenericBlobMetadata::from)
+              .collect(ImmutableList.toImmutableList()));
+    }
+  }
+
+  private List<Blob> generateNDVBlobs() {
+    return NDVSketchGenerator.generateNDVSketchesAndBlobs(spark(), table, snapshot, columns());
+  }
+
+  @Override
+  public ComputeTableStats columns(String... newColumns) {
+    Preconditions.checkArgument(
+        newColumns != null && newColumns.length > 0, "Columns cannot be null/empty");
+    this.columns = ImmutableSet.copyOf(newColumns);
+    return this;
+  }
+
+  @Override
+  public ComputeTableStats snapshot(long newSnapshotId) {
+    Snapshot newSnapshot = table.snapshot(newSnapshotId);
+    Preconditions.checkArgument(newSnapshot != null, "Snapshot not found: %s", newSnapshotId);
+    this.snapshot = newSnapshot;
+    return this;
+  }
+
+  private Set<String> columns() {
+    Schema schema = table.schemas().get(snapshot.schemaId());
+    if (columns == null) {
+      columns = schema.columns().stream().map(Types.NestedField::name).collect(Collectors.toSet());
+    }
+    validateColumns(schema);
+    return columns;
+  }
+
+  private void validateColumns(Schema schema) {
+    for (String columnName : columns) {
+      Types.NestedField field = schema.findField(columnName);
+      if (field == null) {
```

Comment:
What about using `Preconditions` here and adding a bit more context to the error messages?

```
Preconditions.checkArgument(field != null, "Can't find column %s in %s", columnName, schema);
Preconditions.checkArgument(
    field.type().isPrimitiveType(),
    "Can't compute stats on non-primitive type column: %s (%s)",
    columnName,
    field.type());
```